src/app/codegen-data-model/CodegenDataModel_Read.cpp

```diff
@@ -0,0 +1,350 @@
+    case ZCL_INT56U_ATTRIBUTE_TYPE: // Unsigned 56-bit integer
+        return EncodeFromSpan<OddSizedInteger<7, false>>(data, isNullable, encoder);
+    case ZCL_INT64U_ATTRIBUTE_TYPE: // Unsigned 64-bit integer
+        return EncodeFromSpan<uint64_t>(data, isNullable, encoder);
+    case ZCL_INT8S_ATTRIBUTE_TYPE: // Signed 8-bit integer
+        return EncodeFromSpan<int8_t>(data, isNullable, encoder);
+    case ZCL_INT16S_ATTRIBUTE_TYPE: // Signed 16-bit integer
+        return EncodeFromSpan<int16_t>(data, isNullable, encoder);
+    case ZCL_INT24S_ATTRIBUTE_TYPE: // Signed 24-bit integer
+        return EncodeFromSpan<OddSizedInteger<3, true>>(data, isNullable, encoder);
+    case ZCL_INT32S_ATTRIBUTE_TYPE: // Signed 32-bit integer
+        return EncodeFromSpan<int32_t>(data, isNullable, encoder);
+    case ZCL_INT40S_ATTRIBUTE_TYPE: // Signed 40-bit integer
+        return EncodeFromSpan<OddSizedInteger<5, true>>(data, isNullable, encoder);
+    case ZCL_INT48S_ATTRIBUTE_TYPE: // Signed 48-bit integer
+        return EncodeFromSpan<OddSizedInteger<6, true>>(data, isNullable, encoder);
+    case ZCL_INT56S_ATTRIBUTE_TYPE: // Signed 56-bit integer
+        return EncodeFromSpan<OddSizedInteger<7, true>>(data, isNullable, encoder);
+    case ZCL_INT64S_ATTRIBUTE_TYPE: // Signed 64-bit integer
+        return EncodeFromSpan<int64_t>(data, isNullable, encoder);
+    case ZCL_SINGLE_ATTRIBUTE_TYPE: // 32-bit float
+        return EncodeFromSpan<float>(data, isNullable, encoder);
+    case ZCL_DOUBLE_ATTRIBUTE_TYPE: // 64-bit float
+        return EncodeFromSpan<double>(data, isNullable, encoder);
+    case ZCL_CHAR_STRING_ATTRIBUTE_TYPE: // Char string
+        return EncodeStringLike<CharSpan, ShortPascalString>(data, isNullable, encoder);
+    case ZCL_LONG_CHAR_STRING_ATTRIBUTE_TYPE:
+        return EncodeStringLike<CharSpan, LongPascalString>(data, isNullable, encoder);
+    case ZCL_OCTET_STRING_ATTRIBUTE_TYPE: // Octet string
+        return EncodeStringLike<ByteSpan, ShortPascalString>(data, isNullable, encoder);
+    case ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE:
+        return EncodeStringLike<ByteSpan, LongPascalString>(data, isNullable, encoder);
+    default:
+        ChipLogError(DataManagement, "Attribute type 0x%x not handled", static_cast<int>(metadata->attributeType));
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedRead);
+    }
+}
+
+} // namespace
+
+/// separated-out ReadAttribute implementation (given existing complexity)
+///
+/// Generally will:
+///    - validate ACL (only for non-internal requests)
+///    - Try to read attribute via the AttributeAccessInterface
+///    - Try to read the value from ember RAM storage
+CHIP_ERROR CodegenDataModel::ReadAttribute(const InteractionModel::ReadAttributeRequest & request, AttributeValueEncoder & encoder)
+{
+    ChipLogDetail(DataManagement,
+                  "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=%x AttributeId=" ChipLogFormatMEI " (expanded=%d)",
```

Comment:
```suggestion
                  "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=0x%x AttributeId=" ChipLogFormatMEI " (expanded=%d)",
```

Or just %d; usually endpoint numbers are not hex.  I know you just copied this from ReadSingleClusterData...